### ISSUE TITLE
add MTA-STS support

### ISSUE THREAD

**Issue by shoulders** [virtualmin/virtualmin-gpl#808]
|  SYSTEM INFORMATION||
|----------------------|---------------------------|
|  OS type and version |     Ubuntu Linux 22.04.4     |
|  Webmin version      |     2.111      |
|  Usermin version     |     2.010     |
|  Virtualmin version  |     7.10.0       |
|  Theme version       |     21.10      |
|  Package updates     |      30  package updates are available   |

### The background

MTA-STS is an email technology that ensures that there is a secure/encrypted connection between both email servers.

### the request

Can this be added to virtualmin?

### additional
**How Do I Set Up MTA-STS?**

- To implement MTA-STS for your domain(s), you must do the following:
  - Add an A or CNAME type DNS record at mta-sts.[domain] that points to the HTTPS enabled web server serving the MTA-STS policy file.
  - Add a TXT or CNAME type DNS record at _mta-sts.[domain] that indicates the use of MTA-STS, and update the ID value on policy change.
  - Set up an HTTPS enabled web server with a valid certificate for the domain at mta-sts.
  - (Optional/Recommended) Enable SMTP TLS reporting through TXT or CNAME record held at _smtp._tls.domain. Set the Mode to Testing.

from: https://mxtoolbox.com/dmarc/details/mta-sts/what-is-mta-sts-record

### Links

- https://mxtoolbox.com/dmarc/details/mta-sts/what-is-mta-sts-record
- https://mxtoolbox.com/mta-sts.aspx -  a free lookup tool
- https://www.gov.uk/government/publications/email-security-standards/using-the-mail-transfer-agent-strict-transport-security-mta-sts-protocol-in-your-organisation
- https://www.youtube.com/watch?v=3bzzLsQe-Sg - A webinar record for MTS-STS by Synametrics Technologies.
- https://www.youtube.com/watch?v=2c1D_YG_-0w - What is MTA-STS ? (2024)

**Comment by chris001** [comment]
Adding automatic DNS records for MTA Strict Transport Security and TLS Reporting, for all incoming and outgoing postifx MTA emails, would protect users from network eavesdropping by ISPs and data collection organizations.  Very nice to have.  This feature would place Virtualmin ahead of Plesk and cPanel - neither of which have MTA-STS and TLS Reporting (yet).

[Increase email security with MTA-STS and TLS Reporting](https://support.google.com/a/answer/9261504).

**Comment by jcameron** [comment]
Interesting, I didn't know about this ... I'll look into adding support for MTA-STS.

**Comment by iliajie** [comment]
As far as I know, it's not that simple to set up! The MTA-STS policy file must be accessible via HTTPS at a specific URL like:

```
https://mta-sts.yourdomain.com/.well-known/mta-sts.txt
```

This means you need to support DNS records for domains like `mta-sts.`, have an LE SSL certificate set up, and configure Postfix with the `smtp_tls_policy_maps` table.

That's a lot of work! Why not just provide an option in Virtualmin to enforce TLS connections using:

```
smtp_tls_security_level = enforce
```

and

```
smtpd_tls_security_level = enforce
```

**Comment by shoulders** [comment]
- enforcing TLS connection is not RFC compliant, this is why MTA-STS was developed

My thoughts for solution on virtualmin (but not tested) for enabling MTA-STS
- create an alias for the domain, keep it hidden in the GUI but not DNS (obviously)
- deploy the text file into the .well-known directory of the domain
- create a LE SSL for the alias which could be `part` of the main domain certificate

This should prevent the need for setting up another virtual server requiring extra assets. I did not see anywhere that the SSL had to be separate to that of the domain you wish to add MTA-STS to.

**Comment by iliajie** [comment]
Opportunistic TLS, i.e., `smtp*_tls_security_level = may`, is compliant and we're already using it. When I hosted my own mail, I used `enforce` and couldn't care less. It also helped prevent spam. While I understand that using `encrypt` blindly isn't suitable for everyone, especially companies, it's much less of a headache and easier to support. MTA-STS, on the other hand, seems overly complex to support and will likely be treated as a niche feature..

**Comment by shoulders** [comment]
It is niche at the moment, but maybe in the future it will be used more.

I did not know about it either until a couple of days ago when looking at forcing tls only email.

**Comment by chris001** [comment]
Here's some more reasons why it's a good idea to add, followed by a real world risk it protects against.

It's standard now since some time, because it publishes the mail server's settings to the internet, so all mail servers will know what to expect and what is likely an attack.  This allows MTAs to verify email messages are coming from, and going to, the authorized mail server for the domain, and to reject email that is coming from unauthorized spoofed servers.  

Microsoft, Yahoo, Google, Proton mail, all the major email services, are enforcing the MTA-STS policies set by domains sending mail to them and receiving mail from them.

MTA Strict Transport Security was introduced by email service providers as a solution to protect against downgrade and man-in-the-middle attacks in SMTP sessions, as well as solving the lack of security-first communication standards in email.

MTA-STS on its most basic level elevates the security of inbound email being sent to your domains by ensuring 1) only messages that are TLS encrypted are delivered and 2) those messages can only be delivered to those email servers published by the MTA-STS policy.

MTA-STS standardizes allowing domain owners to specify how mail servers should handle the encryption and authentication of their SMTP connections.  MTA-STS can help prevent email spoofing, interception, and tampering by enforcing TLS encryption and certificate validation for your domains - by requiring authentication checks and good encryption for email in transit.

MTA-STS also increases email security by acting as a deterrent against pervasive monitoring of email traffic and protecting against man-in-the-middle attacks.  You can make your email communications more secure by setting MTA-STS policies and ask the organizations with which you communicate to also set MTA-STS policies for their mail servers.

This solves real cyber security problems.  For example: Scammers have already hacked into email, intercepted business invoices, changed the receiving bank account details in the invoice to details of a bank account under their control, and ended up in possession of the business funds, all because of the easy to eavesdrop unencrypted email traffic, and lack of standard to control which named mail servers, verified by TLS certificates, and which encryption protocols, are authorized to send and receive mail for the domain.

**Comment by Neustradamus** [comment]
It is a good request :)

**Comment by shoulders** [comment]
TLS reporting could be added as a Pro feature if this is implemented.

**Comment by verne-work** [comment]
researching MTA-STS today as a client asked about it.

I must be missing something big here ... DNS MX tells the world where I wish my incoming mail to come to.    Requiring starttls in postfix with `smtpd_tls_security_level = encrypt` forces it on (I think thats the correct parameter).   Sure the reporting aspect of MTA-STS sounds nice, but how many senders will actually generate a report (just like SPF and DMARC, its all optional as the sender can ignore parts of it or the whole thing)  😄 

and of course there is the issue of -- how many senders (of incoming mail to my systems) support and honor MTA-STS ?

and what about the local impact and actions -- I am confused about that as well -- will postfix or sendmail look for the local MTA-STS settings and do the right thing?  the only significant item I can see (besides reporting) to MTA-STS is the local cert presented to initiate starttls is validated.

what exactly was MTA-STS trying to fix anyway?

**Comment by shoulders** [comment]
I think it prevents tls downgrading to non encrypted while keeping the non encrypted channel open of a mts record is not specified

**Comment by Lawkss** [comment]
I implemented this today. 

Created subdomain server `mta-sts.mydomain.tld` create a file `mta-sts.txt` in `/.well-known/` with something likes this:

```
version: STSv1
mode: enforce 
mx: *.mydomain.tld 
max_age: 86400
```

Then add two DNS records:

`_mta-sts.mydomain.tld.	IN	TXT	"v=STSv1; id=20250531082020"
_smtp._tls.mydomain.tld. IN TXT "v=TLSRPTv1; rua=mailto:tlsreports@mydomain.tld"`

Fairly simple. 

However, I did not touch postfix, it works with dane, my understanding is it does not interfere without the mta-sts resolver and policy maps.

From RFC 8461:

> The primary motivation of MTA-STS is to provide a mechanism for
>    domains to ensure transport security even when deploying DNSSEC is
>    undesirable or impractical.  However, MTA-STS is designed not to
>    interfere with DANE deployments when the two overlap; in particular,
>    senders who implement MTA-STS validation MUST NOT allow MTA-STS
>    Policy validation to override a failing DANE validation.

Anyway so it does not honor it for outgoing.....

**Comment by Lawkss** [comment]
Actually there seems to be a resolver where Dane is respected

https://github.com/Zuplu/postfix-tlspol

**Comment by jcameron** [comment]
Ok I will start work on MTS STS support shortly ..

**Comment by jcameron** [comment]
Ok this has been implemented for inclusion in the next Virtualmin release!

**Comment by shoulders** [comment]
Nice one pal, I think this will be a popular thing. Next BIMI   lol

**Comment by iliaross** [comment]
> Ok this has been implemented for inclusion in the next Virtualmin release!

Jamie, I believe it's still missing support for the Virtualmin Nginx plugin?

**Comment by jcameron** [comment]
No, Nginx support is implemented (but will still need a new release of the plugin)

**Comment by jcameron** [comment]
See https://github.com/virtualmin/virtualmin-nginx/commit/247d8611babe82b36b1182830116a4a67dff2b5e and https://github.com/virtualmin/virtualmin-gpl/commit/a39432281be9df68761bf1a61e48fa5c7168329f

**Comment by iliaross** [comment]
> See [virtualmin/virtualmin-nginx@247d861](https://github.com/virtualmin/virtualmin-nginx/commit/247d8611babe82b36b1182830116a4a67dff2b5e) and [a394322](https://github.com/virtualmin/virtualmin-gpl/commit/a39432281be9df68761bf1a61e48fa5c7168329f)

Ah, I see the bug now. Check this patch out https://github.com/virtualmin/virtualmin-gpl/commit/c82fecd9850d033efdb5469b94c34eccb18e2ec9.
